code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm

```diff
@@ -0,0 +1,529 @@
+			tlv.set_value(threshold_type, value)
+			investigate_log("threshold value for [threshold]:[threshold_type] was set to [value] by [key_name(usr)]", INVESTIGATE_ATMOS)
+
+			var/turf/our_turf = get_turf(src)
+			var/datum/gas_mixture/environment = our_turf.return_air()
+			check_air_dangerlevel(our_turf, environment, environment.temperature)
+
+		if("reset_threshold")
+			var/threshold = text2path(params["threshold"]) || params["threshold"]
+			var/datum/tlv/tlv = tlv_collection[threshold]
+			if(isnull(tlv))
+				return
+			var/threshold_type = params["threshold_type"]
+			tlv.reset_value(threshold_type)
+			investigate_log("threshold value for [threshold]:[threshold_type] was reset by [key_name(usr)]", INVESTIGATE_ATMOS)
+
+			var/turf/our_turf = get_turf(src)
+			var/datum/gas_mixture/environment = our_turf.return_air()
+			check_air_dangerlevel(our_turf, environment, environment.temperature)
+
+		if ("alarm")
+			if (alarm_manager.send_alarm(ALARM_ATMOS))
+				danger_level = AIR_ALARM_ALERT_SEVERE
+
+		if ("reset")
+			if (alarm_manager.clear_alarm(ALARM_ATMOS))
+				danger_level = AIR_ALARM_ALERT_NONE
+
+	update_appearance()
+
+	return TRUE
+
+/obj/machinery/airalarm/update_appearance(updates)
+	. = ..()
+
+	if(panel_open || (machine_stat & (NOPOWER|BROKEN)) || shorted)
+		set_light(0)
+		return
+
+	var/area/our_area = get_area(src)
+	var/color
+	switch(max(danger_level, !!our_area.active_alarms[ALARM_ATMOS]))
+		if(0)
+			color = "#03A728" // green
+		if(1)
+			color = "#EC8B2F" // yellow
+		if(2)
+			color = "#DA0205" // red
+
+	set_light(1.4, 1, color)
+
+/obj/machinery/airalarm/update_icon_state()
+	if(panel_open)
+		switch(buildstage)
+			if(AIRALARM_BUILD_COMPLETE)
+				icon_state = "alarmx"
+			if(AIRALARM_BUILD_NO_WIRES)
+				icon_state = "alarm_b2"
+			if(AIRALARM_BUILD_NO_CIRCUIT)
+				icon_state = "alarm_b1"
+		return ..()
+
+	icon_state = "alarmp"
+	return ..()
+
+/obj/machinery/airalarm/update_overlays()
+	. = ..()
+
+	if(panel_open || (machine_stat & (NOPOWER|BROKEN)) || shorted)
+		return
+
+	var/area/our_area = get_area(src)
+	var/state
+	switch(max(danger_level, !!our_area.active_alarms[ALARM_ATMOS]))
+		if(0)
+			state = "alarm0"
+		if(1)
+			state = "alarm2" //yes, alarm2 is yellow alarm
+		if(2)
+			state = "alarm1"
+
+	. += mutable_appearance(icon, state)
+	. += emissive_appearance(icon, state, src, alpha = src.alpha)
+
+/**
+ * main proc for throwing a shitfit if the air isnt right.
+ * goes into warning mode if gas parameters are beyond the tlv warning bounds, goes into hazard mode if gas parameters are beyond tlv hazard bounds
+ *
+ */
+/obj/machinery/airalarm/proc/check_air_dangerlevel(turf/location, datum/gas_mixture/environment, exposed_temperature)
+	SIGNAL_HANDLER
+	if((machine_stat & (NOPOWER|BROKEN)) || shorted)
+		return
+
+	var/datum/tlv/current_tlv
+	//cache for sanic speed (lists are references anyways)
+	var/list/cached_tlv = tlv_collection
+
+	var/list/env_gases = environment.gases
```

Comment:
move this above its loop, shouldn't stick it here (I get for the comment, hm s annoying)